planner/core/logical_plan_builder.go

```diff
@@ -2962,62 +2982,139 @@ func (b *PlanBuilder) buildWindowFunctionFrame(spec *ast.WindowSpec, orderByItem
 	//   (1) With order by, the default frame is equivalent to "RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW";
 	//   (2) Without order by, the default frame is equivalent to "RANGE BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING",
 	//       which is the same as an empty frame.
-	if needFrame && expr.Spec.Frame == nil && len(orderBy) > 0 {
-		expr.Spec.Frame = &ast.FrameClause{
+	if needFrame && spec.Frame == nil && spec.OrderBy != nil {
+		newSpec := *spec
+		newSpec.Frame = &ast.FrameClause{
 			Type: ast.Ranges,
 			Extent: ast.FrameExtent{
 				Start: ast.FrameBound{Type: ast.Preceding, UnBounded: true},
 				End:   ast.FrameBound{Type: ast.CurrentRow},
 			},
 		}
+		return &newSpec, true
 	}
 	// For functions that operate on the entire partition, the frame clause will be ignored.
-	if !needFrame && expr.Spec.Frame != nil {
-		b.ctx.GetSessionVars().StmtCtx.AppendNote(ErrWindowFunctionIgnoresFrame.GenWithStackByArgs(expr.F, expr.Spec.Name.O))
-		expr.Spec.Frame = nil
-	}
-	frame, err := b.buildWindowFunctionFrame(&expr.Spec, orderBy)
-	if err != nil {
-		return nil, err
+	if !needFrame && spec.Frame != nil {
+		specName := spec.Name.O
+		b.ctx.GetSessionVars().StmtCtx.AppendNote(ErrWindowFunctionIgnoresFrame.GenWithStackByArgs(name, specName))
+		newSpec := *spec
+		newSpec.Frame = nil
+		return &newSpec, true
+	}
+	return spec, false
+}
+
+// groupWindowFuncs groups the window functions according to the window specification name.
+// TODO: We can group the window function by the definition of window specification.
+func (b *PlanBuilder) groupWindowFuncs(windowFuncs []*ast.WindowFuncExpr) (map[*ast.WindowSpec][]*ast.WindowFuncExpr, error) {
+	// updatedSpec is used to handle the specifications that have frame clause changed.
+	updatedSpec := make(map[string]*ast.WindowSpec)
+	groupedWindow := make(map[*ast.WindowSpec][]*ast.WindowFuncExpr)
+	for _, windowFunc := range windowFuncs {
+		if windowFunc.Spec.Name.L == "" {
+			windowFunc.Spec.Name = model.NewCIStr("<unnamed window>")
+			spec := &windowFunc.Spec
+			if spec.Ref.L != "" {
+				ref, ok := b.windowSpecs[spec.Ref.L]
+				if !ok {
+					return nil, ErrWindowNoSuchWindow.GenWithStackByArgs(spec.Ref.O)
+				}
+				err := mergeWindowSpec(spec, ref)
+				if err != nil {
+					return nil, err
+				}
+			}
+			spec, _ = b.handleDefaultFrame(spec, windowFunc.F)
+			groupedWindow[spec] = append(groupedWindow[spec], windowFunc)
+			continue
+		}
+
+		name := windowFunc.Spec.Name.L
+		spec, ok := b.windowSpecs[name]
+		if !ok {
+			return nil, ErrWindowNoSuchWindow.GenWithStackByArgs(windowFunc.Spec.Name.O)
+		}
+		newSpec, updated := b.handleDefaultFrame(spec, windowFunc.F)
+		if !updated {
+			groupedWindow[spec] = append(groupedWindow[spec], windowFunc)
+		} else {
+			if _, ok := updatedSpec[name]; !ok {
+				updatedSpec[name] = newSpec
+			}
+			updatedSpec := updatedSpec[name]
```

Comment:
change the map `updatedSpec` to `updatedSpecMap` or something else to avoid the same name?